### DISCUSSION TITLE
Passwordless authentication

### DISCUSSION THREAD

**Discussion by Para-lyzed** [dani-garcia/vaultwarden#2817] [Ideas]
Bitwarden official has just implemented a system that allows for passwordless authentication through verification from the mobile app. I am not sure how much work would need to be done to implement this in Vaultwarden, but I wanted to bring it up as a potential (probably low priority) feature.

**Comment by BlackDex** [comment]
I'm afraid that isn't something we can add. A quick look at the code looks like it is linked to push notifications which Vaultwarden doesn't support.

**Reply by Para-lyzed** [reply]
Got it, makes sense.

**Reply by tessus** [reply]
@BlackDex I thought vaultwarden supported push notifications via websockets. When it comes to the mobile apps, I guess additional code would have to be added to the server, is that correct?

e.g. there is the [notify_push](https://github.com/nextcloud/notify_push) for Nextcloud that is written in Rust. maybe some of that code can be re-used. What do you think?

**Reply by BlackDex** [reply]
Push notifications do not work via websockets.
The websockets only work for the extension, web and desktop clients as far as i know.

Mobile clients use the services from Google or Apple. The f-droid version for example doesn't have push notifications at all since it laks the Google services integration.

Would be nice if they have websockets support though.

**Reply by tessus** [reply]
Ok, but what has to be done so that push notifications work on mobile? Since the official bitwarden clients obviously have these Google/Apple services included, what has to be done on the vaultwarden server? (The clients even allow a customized "Notifications Server URL" to be set.)

I am just curious as to what is necessary to get this working.

**Reply by BlackDex** [reply]
As far as i understand the communication all goes via the Bitwarden servers. You also need to have a Bitwarden license for a self hosted environment, else this won't work.

What happens is the self-hosted installation send a call using the license, server id and client id to a Bitwarden server which in turn triggers the Google or Apple push services to send a message.

Also see https://github.com/dani-garcia/vaultwarden/issues/126#issuecomment-412103636

And also a few comments here https://github.com/dani-garcia/vaultwarden/issues/246#issuecomment-475879502

**Reply by tessus** [reply]
Yep, I read these comments but I don't understand them, because there are other servers that can do this. Why does it have to be a Bitwarden server? You cloned pretty much every other functionality. I also sent a link how Nextcloud (via the client push app) sends push notifications to their mobile clients. It is written in Rust, so maybe this can be used.

**Reply by BlackDex** [reply]
And how are we going to push to the clients then?? There need to be client support for this. The clients currently register to Google firebase for example to receive push notifications. We can't access that firebase from Vaultwarden, that is where push.bitwarden.com is for.

And we can't access push.bitwarden.com because Vaultwarden has no server id and license key provided by Bitwarden.

Looking at the nextcloud link you provided, that all works via websockets. The clients keep a websocket connection open to receive these notifications. That functionality is not build in into the Bitwarden mobile clients.

So, if you want this to be workable for Vaultwarden then Bitwarden needs to get websocket support for there mobile clients.

**Reply by tessus** [reply]
Thanks for the explanation. 

I think it makes a bit more sense now. That is if `push.bitwarden.com` is really hardcoded. I [couldn't find it](https://github.com/bitwarden/clients/search?q=push.bitwarden.com) in the source code. 

> So, if you want this to be workable for Vaultwarden then Bitwarden needs to get websocket support for there mobile clients.

In this case, I hope and wait. ;-) Although I doubt that they will do that since they alraedy have a system that works for them.

**Reply by BlackDex** [reply]
The clients do not talk with `push.bitwarden.com`, only the server(s) do. The mobile clients connect to the push services provided by Google and Apple. On a self-hosted server, during a change, it sends a web-socket message to the web-socket clients, and when configured it will also send a push message through the push relay provided by Bitwarden, which in turn triggers the clients.

* Android: https://github.com/bitwarden/mobile/tree/master/src/Android/Push
* iOS: https://github.com/bitwarden/mobile/tree/master/src/iOS/Services

All other stuff is only on the server code base, since all other clients do not use that push services, it uses web-sockets.
For the server stuff see here:
* https://github.com/bitwarden/server/blob/0ea70f68f13adc5e9bb9abc26bf4b6773dd30791/src/Core/Services/Implementations/RelayPushRegistrationService.cs#L11-L24
* https://github.com/bitwarden/server/blob/7f5f010e1eea400300c47f776604ecf46c4b4f2d/util/Setup/Configuration.cs#L86-L88

The only way for this to work is to have the mobile clients support web-sockets, which in theory is possible of course.
And maybe it would even be a good idea for the F-Droid versions as an alternative, because those do not support push notifications either.

**Reply by jjlin** [reply]
I doubt Bitwarden would ever add support for WebSockets on mobile, as they probably have very few users using F-Droid (relatively), so it has little benefit for the work involved. If they added a passwordless login method that involved scanning a QR code on your phone, that could be more readily implemented.

**Reply by BlackDex** [reply]
> If they added a passwordless login method that involved scanning a QR code on your phone, that could be more readily implemented.

Which i doubt they would add, since they really want to use biometrics for this. Since a QR is something a hacker could steal remotely for example.

**Reply by Tealk** [reply]
Wouldn't [unifiedpush](https://unifiedpush.org/) be a good solution?

https://docs.ntfy.sh/subscribe/phone/#unifiedpush

**Reply by BlackDex** [reply]
@Tealk unfortunately not. Looks like the mobile app needs support for it. I actually think that websockets would be fine for this. But that also needs to be built in. It could help there F-droid user base, but not sure that is big enough.

**Reply by nWacky** [reply]
It seems like mobile app [sends FCM token to the backend in `/connect/token`](https://github.com/bitwarden/mobile/blob/9ae269dd570f69ebd10c0a4752d0b0f4b88c1bb7/src/Core/Services/ApiService.cs#L90-L93). Then the server [has `devicePushToken`](https://github.com/bitwarden/server/blob/e340cba8fcd3c6d4aebe96d9e9ca2a99f8281059/src/Identity/IdentityServer/BaseRequestValidator.cs#L395-L401)

Firebase doesn't allow sending notifications without proper client/server credentials pair, so the bitwarden server [is using push.bitwarden.com relay](https://github.com/bitwarden/server/blob/0bcd8d0b411a81a8df5c5939fee603957213290f/src/Core/Services/Implementations/RelayPushNotificationService.cs#L193-L201)

Bitwarden has [client credentials](https://bitwarden.com/help/oidc-okta/#get-client-credentials) for self hosted bitwarden servers. Perhaps, we could use push.bitwarden.com  and somehow pass in client credentials from an env variable?

**Reply by BlackDex** [reply]
This has been discussed a few times already. We don't want to use  there push service without some kind of approval. Besides that. It will make usage of there service which could have impact on there system. And we don't want that either.

**Comment by ghost** [comment]
@BlackDex Good news at least it's on their roadmap one post from them on a poll about this feature : https://fosstodon.org/@bitwarden/109558587626234859

**Comment by Tyree** [comment]
Ah man. This is a bummer. I just saw the blog post from December saying "Login with Device" was supported and I got all excited! :-D
Thanks for the hard work, guys!

**Comment by quexten** [comment]
Since mobile-push notifications are now merged in master, I made a WIP Pull-request:

https://github.com/dani-garcia/vaultwarden/pull/3592
